src/py21cmfast/wrapper/cfuncs.py

```diff
@@ -366,931 +434,393 @@ def construct_fftw_wisdoms(
-        if ap_c["ALPHA_STAR_MINI"]
-        else 0.0
-    )
-    growthf = lib.dicke(redshift)
-    t_h = (1 / cosmo_params.cosmo.H(redshift)).to("s").value
-    # Unfortunately we have to do this until we sort out the USE_INTERPOLATION_TABLES flag
-    # Since these integrals take forever if the flag is false
-    if return_integral:
-        Xray = np.vectorize(lib.Xray_ConditionalM)(
-            redshift,
-            growthf,
-            np.log(M_min),
-            np.log(M_max),
-            np.log(cond_mass),
-            sigma_cond,
-            densities[:, None] if flag_options.USE_MINI_HALOS else densities,
-            mcrit_atom,
-            10 ** l10mturns[None, :] if flag_options.USE_MINI_HALOS else 1.0,
-            ap_c["ALPHA_STAR"],
-            ap_c["ALPHA_STAR_MINI"],
-            ap_c["F_STAR10"],
-            ap_c["F_STAR7_MINI"],
-            mlim_fstar_acg,
-            mlim_fstar_mcg,
-            ap_c["L_X"],
-            ap_c["L_X_MINI"],
-            t_h,
-            ap_c["t_STAR"],
-            user_params.cdict["INTEGRATION_METHOD_ATOMIC"],
-        )
-        return Xray
-
-    # TODO: this function needs cleanup
-    if user_params.USE_INTERPOLATION_TABLES:
-        lib.initialise_Xray_Conditional_table(
-            densities.min() - 0.01,
-            densities.max() + 0.01,
-            redshift,
-            M_min,
-            M_max,
-            cond_mass,
-            ap_c["ALPHA_STAR"],
-            ap_c["ALPHA_STAR_MINI"],
-            ap_c["F_STAR10"],
-            ap_c["F_STAR7_MINI"],
-            ap_c["L_X"],
-            ap_c["L_X_MINI"],
-            t_h,
-            ap_c["t_STAR"],
+    """Evaluates the conditional star formation rate density expected at a range of densities."""
+    if densities.shape != log10mturns.shape:
+        raise ValueError(
+            "the shapes of the input arrays `cond_array` and `probabilities"
+            " must be equal."
         )
 
-    Xray = np.vectorize(lib.EvaluateXray_Conditional)(
-        densities[:, None] if flag_options.USE_MINI_HALOS else densities,
-        l10mturns[None, :] if flag_options.USE_MINI_HALOS else mcrit_atom,
+    densities = densities.astype("f8")
+    log10mturns = log10mturns.astype("f8")
+    xray = np.zeros_like(densities)
+
+    lib.get_conditional_Xray(
+        inputs.user_params.cstruct,
+        inputs.cosmo_params.cstruct,
+        inputs.astro_params.cstruct,
+        inputs.flag_options.cstruct,
         redshift,
-        growthf,
-        M_min,
-        M_max,
-        cond_mass,
-        sigma_cond,
-        mcrit_atom,
-        t_h,
-        mlim_fstar_acg,
-        mlim_fstar_mcg,
+        radius,
+        densities.size,
+        ffi.cast("double *", ffi.from_buffer(densities)),
+        ffi.cast("double *", ffi.from_buffer(log10mturns)),
+        ffi.cast("double *", ffi.from_buffer(xray)),
     )
-    return Xray
+
+    return xray
 
 
 def halo_sample_test(
```

Comment:
I'm a little confused with the name of this function -- is it purely a function for testing? If so, should it be defined in this module or in the tests themselves?